File: java/src/test/java/org/mydotey/scf/TestDynamicConfigurationSource.java
package org.mydotey.scf;

import java.util.HashMap;
import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;

import org.mydotey.scf.ConfigurationSourceConfig;

/**
 * @author koqizhao
 *
 * May 17, 2018
 */
public class TestDynamicConfigurationSource extends TestConfigurationSource {

    public TestDynamicConfigurationSource(ConfigurationSourceConfig config, HashMap<String, String> properties) {
        super(config, properties);
    }

    @Override
    protected void init() {
        _properties = new ConcurrentHashMap<>();
    }

    public void setPropertyValue(String key, String value) {
        String oldValue = _properties.get(key);
        if (Objects.equals(oldValue, value))
            return;

        if (value == null)
            _properties.remove(key);
        else
            _properties.put(key, value);

        raiseChangeEvent();
    }

}

File: dotnet/test/MyDotey.SCF.Tests/TestDynamicConfigurationSource.cs
using System;
using System.Collections.Generic;
using System.Collections.Concurrent;

namespace MyDotey.SCF
{
    /**
     * @author koqizhao
     *
     * May 17, 2018
     */
    public class TestDynamicConfigurationSource : TestConfigurationSource
    {
        public TestDynamicConfigurationSource(ConfigurationSourceConfig config, Dictionary<String, String> properties)

            : base(config, properties)
        {
        }

        protected override void Init()
        {
            _properties = new ConcurrentDictionary<String, String>();
        }

        public virtual void SetPropertyValue(String key, String value)
        {
            _properties.TryGetValue(key, out String oldValue);
            if (Object.Equals(oldValue, value))
                return;

            if (value == null)
                _properties.Remove(key);
            else
                _properties[key] = value;

            RaiseChangeEvent();
        }
    }
}